// @typegoose/typegoose/lib/globalOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const data_1 = require("./internal/data");
const logSettings_1 = require("./logSettings");
const types_1 = require("./types");
/**
 * Set Typegoose's global Options
 */
function setGlobalOptions(options) {
    if (typeof options !== 'object') {
        throw new TypeError('"options" argument needs to be an object!');
    }
    logSettings_1.logger.info('"setGlobalOptions" got called with', options);
    for (const key of Object.keys(options)) {
        data_1.globalOptions[key] = Object.assign({}, data_1.globalOptions[key], options[key]);
    }
    logSettings_1.logger.info('new Global Options:', options);
}
exports.setGlobalOptions = setGlobalOptions;
/**
 * Parse Typegoose Environment Variables and apply them
 */
function parseENV() {
    var _a;
    logSettings_1.logger.info('"parseENV" got called');
    if (((_a = process.env.TG_USE_NEW_ENUM) === null || _a === void 0 ? void 0 : _a.length) > 0) {
        logSettings_1.logger.warn('TG_USE_NEW_ENUM & useNewEnum got deprecated, see changelog 6.2 for more');
    }
    const options = {
        globalOptions: {},
        options: {
            allowMixed: process.env.TG_ALLOW_MIXED in types_1.Severity ? mapValueToSeverity(process.env.TG_ALLOW_MIXED) : data_1.globalOptions.options.allowMixed
        }
    };
    setGlobalOptions(options);
}
exports.parseENV = parseENV;
/**
 * Maps strings to the number
 * -> This function is specifically build for "Severity"-Enum
 * @throws {Error} if not in range of the "Severity"-Enum
 * @example
 * ```ts
 * mapValueToSeverity("WARN") === 1
 * mapValueToSeverity("1") === 1
 * // now internal use
 * mapValueToSeverity(1) === 1
 * ```
 * @param value The value to check for
 */
function mapValueToSeverity(value) {
    if (!(value in types_1.Severity)) {
        throw new Error(`"value" is not in range of "Severity"! (got: ${value})`);
    }
    if (typeof value === 'number') {
        return value;
    }
    return mapValueToSeverity(types_1.Severity[value]);
}
