// @typegoose/typegoose/lib/typegoose.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/* imports */
const mongoose = require("mongoose");
exports.mongoose = mongoose;
require("reflect-metadata");
const semver = require("semver");
const util_1 = require("util");
/* istanbul ignore next */
if (semver.lt(mongoose.version, '5.9.2')) {
    throw new Error('Please use mongoose 5.9.2 or higher');
}
if (semver.lt(process.version.slice(1), '8.10.0')) {
    logSettings_1.logger.warn('You are using a NodeJS Version below 8.10.0, Please Upgrade!');
}
const defaultClasses = require("./defaultClasses");
exports.defaultClasses = defaultClasses;
const globalOptions_1 = require("./globalOptions");
exports.setGlobalOptions = globalOptions_1.setGlobalOptions;
const constants_1 = require("./internal/constants");
const data_1 = require("./internal/data");
const errors_1 = require("./internal/errors");
const schema_1 = require("./internal/schema");
const utils_1 = require("./internal/utils");
const logSettings_1 = require("./logSettings");
var logSettings_2 = require("./logSettings");
exports.setLogLevel = logSettings_2.setLogLevel;
exports.LogLevels = logSettings_2.LogLevels;
tslib_1.__exportStar(require("./prop"), exports);
tslib_1.__exportStar(require("./hooks"), exports);
tslib_1.__exportStar(require("./plugin"), exports);
tslib_1.__exportStar(require("."), exports);
tslib_1.__exportStar(require("./typeguards"), exports);
tslib_1.__exportStar(require("./optionsProp"), exports);
var types_1 = require("./types");
exports.Severity = types_1.Severity;
var utils_2 = require("./internal/utils");
exports.getClassForDocument = utils_2.getClassForDocument;
exports.getClass = utils_2.getClass;
globalOptions_1.parseENV(); // call this before anything to ensure they are applied
/** @deprecated */
class Typegoose {
    /* istanbul ignore next */
    constructor() {
        // tslint:disable-next-line:no-empty
        util_1.deprecate(() => { }, 'Typegoose Class is Deprecated!')();
    }
    /* istanbul ignore next */
    /** @deprecated */
    getModelForClass(cl, settings) {
        return util_1.deprecate(getModelForClass.bind(undefined, cl, settings), 'Typegoose Class is Deprecated!');
    }
    /* istanbul ignore next */
    /** @deprecated */
    setModelForClass(cl, settings) {
        return util_1.deprecate(getModelForClass.bind(undefined, cl, settings), 'Typegoose Class is Deprecated!');
    }
    /* istanbul ignore next */
    /** @deprecated */
    buildSchema(cl) {
        return util_1.deprecate(buildSchema.bind(undefined, cl), 'Typegoose Class is Deprecated!');
    }
}
exports.Typegoose = Typegoose;
/**
 * Get a Model for a Class
 * Executes .setModelForClass if it can't find it already
 * @param cl The uninitialized Class
 * @returns The Model
 * @public
 * @example
 * ```ts
 * class Name {}
 *
 * const NameModel = getModelForClass(Name);
 * ```
 */
function getModelForClass(cl, options) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    if (typeof cl !== 'function') {
        throw new errors_1.NoValidClass(cl);
    }
    options = typeof options === 'object' ? options : {};
    const roptions = utils_1.mergeMetadata(constants_1.DecoratorKeys.ModelOptions, options, cl);
    const name = utils_1.getName(cl);
    if (data_1.models.has(name)) {
        return data_1.models.get(name);
    }
    const model = (_f = (_c = (_b = (_a = roptions) === null || _a === void 0 ? void 0 : _a.existingConnection) === null || _b === void 0 ? void 0 : _b.model.bind(roptions.existingConnection), (_c !== null && _c !== void 0 ? _c : (_e = (_d = roptions) === null || _d === void 0 ? void 0 : _d.existingMongoose) === null || _e === void 0 ? void 0 : _e.model.bind(roptions.existingMongoose))), (_f !== null && _f !== void 0 ? _f : mongoose.model.bind(mongoose)));
    const compiledmodel = model(name, buildSchema(cl, roptions.schemaOptions));
    const refetchedOptions = (_g = Reflect.getMetadata(constants_1.DecoratorKeys.ModelOptions, cl), (_g !== null && _g !== void 0 ? _g : {}));
    if ((_j = (_h = refetchedOptions) === null || _h === void 0 ? void 0 : _h.options) === null || _j === void 0 ? void 0 : _j.runSyncIndexes) {
        compiledmodel.syncIndexes();
    }
    return addModelToTypegoose(compiledmodel, cl);
}
exports.getModelForClass = getModelForClass;
/**
 * Get Model from internal cache
 * @param key ModelName key
 */
function getModelWithString(key) {
    if (typeof key !== 'string') {
        throw new TypeError(util_1.format('Expected "key" to be a string, got "%s"', key));
    }
    return data_1.models.get(key);
}
exports.getModelWithString = getModelWithString;
/* istanbul ignore next */
/**
 * Builds the Schema & The Model
 * DEPRECTAED: use getModelForClass
 * @param cl The uninitialized Class
 * @returns The Model
 * @deprecated
 */
function setModelForClass(cl) {
    return util_1.deprecate(getModelForClass.bind(undefined, cl), 'setModelForClass is deprecated, please use getModelForClass (see README#Migrate to 6.0.0)');
}
exports.setModelForClass = setModelForClass;
/**
 * Generates a Mongoose schema out of class props, iterating through all parents
 * @param cl The not initialized Class
 * @returns Returns the Build Schema
 */
function buildSchema(cl, options) {
    var _a;
    if (typeof cl !== 'function') {
        throw new errors_1.NoValidClass(cl);
    }
    const mergedOptions = utils_1.mergeSchemaOptions(options, cl);
    let sch;
    /** Parent Constructor */
    let parentCtor = Object.getPrototypeOf(cl.prototype).constructor;
    // iterate trough all parents
    while (((_a = parentCtor) === null || _a === void 0 ? void 0 : _a.name) !== 'Object') {
        /* istanbul ignore next */
        if (parentCtor.name === 'Typegoose') { // TODO: remove this "if", if the Typegoose class gets removed [DEPRECATION]
            util_1.deprecate(() => undefined, 'The Typegoose Class is deprecated, please try to remove it')();
            break;
        }
        // extend schema
        sch = schema_1._buildSchema(parentCtor, sch, mergedOptions);
        // set next parent
        parentCtor = Object.getPrototypeOf(parentCtor.prototype).constructor;
    }
    // get schema of current model
    sch = schema_1._buildSchema(cl, sch, mergedOptions);
    return sch;
}
exports.buildSchema = buildSchema;
/**
 * This can be used to add custom Models to Typegoose, with the type information of cl
 * Note: no gurantee that the type information is fully correct
 * @param model The model to store
 * @param cl The Class to store
 * @example
 * ```ts
 * class T {}
 *
 * const schema = buildSchema(T);
 * // modifications to the schame can be done
 * const model = addModelToTypegoose(mongoose.model(schema), T);
 * ```
 */
function addModelToTypegoose(model, cl) {
    if (!(model.prototype instanceof mongoose.Model)) {
        throw new TypeError(`"${model}" is not a valid Model!`);
    }
    if (typeof cl !== 'function') {
        throw new errors_1.NoValidClass(cl);
    }
    const name = utils_1.getName(cl);
    if (data_1.models.has(name)) {
        throw new Error(util_1.format('It seems like "addModelToTypegoose" got called twice\n'
            + 'Or multiple classes with the same name are used, which is not supported!'
            + '(%s)', name));
    }
    if (data_1.constructors.get(name)) {
        logSettings_1.logger.info('Class "%s" already existed in the constructors Map', name);
    }
    data_1.models.set(name, model);
    data_1.constructors.set(name, cl);
    return data_1.models.get(name);
}
exports.addModelToTypegoose = addModelToTypegoose;
/**
 * Deletes an existing model so that it can be overwritten
 * with another model
 *
 * @param key
 */
function deleteModel(name) {
    if (typeof name !== 'string') {
        throw new TypeError('name is not an string! (deleteModel)');
    }
    if (!data_1.models.has(name)) {
        throw new Error(`Model "${name}" could not be found`);
    }
    logSettings_1.logger.debug('Deleting Model "%s"', name);
    data_1.models.get(name).db.deleteModel(name);
    data_1.models.delete(name);
    data_1.constructors.delete(name);
}
exports.deleteModel = deleteModel;
/**
 * Delete a model, with the given class
 * @param cl The Class
 */
function deleteModelWithClass(cl) {
    if (typeof cl !== 'function') {
        throw new errors_1.NoValidClass(cl);
    }
    return deleteModel(utils_1.getName(cl));
}
exports.deleteModelWithClass = deleteModelWithClass;
/**
 * Build a Model from a given class and return the model
 * @param from The Model to build From
 * @param cl The Class to make a model out
 * @param id The Identifier to use to differentiate documents (default: cl.name)
 * @example
 * ```ts
 * class C1 {}
 * class C2 extends C1 {}
 *
 * const C1Model = getModelForClass(C1);
 * const C2Model = getDiscriminatorModelForClass(C1Model, C1);
 * ```
 */
function getDiscriminatorModelForClass(from, cl, id) {
    if (!(from.prototype instanceof mongoose.Model)) {
        throw new TypeError(`"${from}" is not a valid Model!`);
    }
    if (typeof cl !== 'function') {
        throw new errors_1.NoValidClass(cl);
    }
    const name = utils_1.getName(cl);
    if (data_1.models.has(name)) {
        return data_1.models.get(name);
    }
    const sch = buildSchema(cl);
    const discriminatorKey = sch.get('discriminatorKey');
    if (sch.path(discriminatorKey)) {
        sch.paths[discriminatorKey].options.$skipDiscriminatorCheck = true;
    }
    const model = from.discriminator(name, sch, id ? id : name);
    return addModelToTypegoose(model, cl);
}
exports.getDiscriminatorModelForClass = getDiscriminatorModelForClass;
